From b5f2cf484b3b4043de51177ac9d8d58e9486c932 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Mon, 15 Apr 2024 10:49:46 +0200
Subject: [PATCH 0459/1695] scpfw: module wide: scmi clock duty cycle support

Implements a clock protocol extension to return a clock duty cycle.
Layers returns a balance 1:1 (50%) duty cycle information when it's not
supported by callee layer.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 1faee53cf405a875c2df4bb71e0bdae09c6df90a scpfw: module wide: scmi clock duty cycle support]
Change-Id: Ifdbb98ec6e8f773d714fffd4d5c38ebec925bf29
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
---
 .../module/clock/include/mod_clock.h          | 30 +++++++++
 .../SCP-firmware/module/clock/src/mod_clock.c | 15 +++++
 .../module/optee/clock/src/mod_optee_clock.c  | 39 ++++++++++++
 .../module/scmi/include/mod_scmi_std.h        |  1 +
 .../scmi_clock/include/internal/scmi_clock.h  | 14 +++++
 .../module/scmi_clock/src/mod_scmi_clock.c    | 61 +++++++++++++++++++
 6 files changed, 160 insertions(+)

diff --git a/core/lib/scmi-server/SCP-firmware/module/clock/include/mod_clock.h b/core/lib/scmi-server/SCP-firmware/module/clock/include/mod_clock.h
index fd8efbb37..deb1d9764 100644
--- a/core/lib/scmi-server/SCP-firmware/module/clock/include/mod_clock.h
+++ b/core/lib/scmi-server/SCP-firmware/module/clock/include/mod_clock.h
@@ -429,6 +429,21 @@ struct mod_clock_drv_api {
         fwk_id_t clock_id,
         uint64_t in_rate,
         uint64_t *out_rate);
+
+    /*!
+     * \brief Get the clock duty cycle.
+     *
+     * \param clock_id Clock device identifier.
+     *
+     * \param[out] num Numerator of the clock duty cycle.
+     *
+     * \param[out] den Denominator of the clock duty cycle.
+     *
+     * \retval ::FWK_SUCCESS The operation succeeded.
+     * \return One of the standard framework error codes.
+     */
+    int (*get_duty_cycle)(fwk_id_t clock_id, uint32_t *num, uint32_t *den);
+
 };
 
 /*!
@@ -543,6 +558,21 @@ struct mod_clock_api {
      * \return One of the standard framework error codes.
      */
     int (*get_info)(fwk_id_t clock_id, struct mod_clock_info *info);
+
+    /*!
+     * \brief Get the clock duty cycle.
+     *
+     * \param clock_id Clock device identifier.
+     *
+     * \param[out] num Numerator of the clock duty cycle.
+     *
+     * \param[out] den Denominator of the clock duty cycle.
+     *
+     * \retval ::FWK_SUCCESS The operation succeeded.
+     * \return One of the standard framework error codes.
+     */
+    int (*get_duty_cycle)(fwk_id_t clock_id, uint32_t *num, uint32_t *den);
+
 };
 
 /*!
diff --git a/core/lib/scmi-server/SCP-firmware/module/clock/src/mod_clock.c b/core/lib/scmi-server/SCP-firmware/module/clock/src/mod_clock.c
index c71202098..2e49a9130 100644
--- a/core/lib/scmi-server/SCP-firmware/module/clock/src/mod_clock.c
+++ b/core/lib/scmi-server/SCP-firmware/module/clock/src/mod_clock.c
@@ -346,6 +346,20 @@ static int clock_get_info(fwk_id_t clock_id, struct mod_clock_info *info)
     return FWK_SUCCESS;
 }
 
+static int clock_get_duty_cycle(fwk_id_t clock_id, uint32_t *num, uint32_t *den)
+{
+    struct clock_dev_ctx *ctx;
+
+    clock_get_ctx(clock_id, &ctx);
+
+    if (!ctx->api->get_duty_cycle) {
+        return FWK_E_SUPPORT;
+    }
+
+    return ctx->api->get_duty_cycle(ctx->config->driver_id, num, den);
+}
+
+
 static const struct mod_clock_api clock_api = {
     .set_rate = clock_set_rate,
     .get_rate = clock_get_rate,
@@ -353,6 +367,7 @@ static const struct mod_clock_api clock_api = {
     .set_state = clock_set_state,
     .get_state = clock_get_state,
     .get_info = clock_get_info,
+    .get_duty_cycle = clock_get_duty_cycle,
 };
 
 /*
diff --git a/core/lib/scmi-server/SCP-firmware/module/optee/clock/src/mod_optee_clock.c b/core/lib/scmi-server/SCP-firmware/module/optee/clock/src/mod_optee_clock.c
index 29254d489..d23d421a3 100644
--- a/core/lib/scmi-server/SCP-firmware/module/optee/clock/src/mod_optee_clock.c
+++ b/core/lib/scmi-server/SCP-firmware/module/optee/clock/src/mod_optee_clock.c
@@ -306,6 +306,44 @@ static int get_rate_from_index(fwk_id_t dev_id,
     return FWK_SUCCESS;
 }
 
+static int get_duty_cycle(fwk_id_t dev_id, uint32_t *num, uint32_t *den)
+{
+    struct optee_clock_dev_ctx *ctx = elt_id_to_ctx(dev_id);
+    struct clk_duty duty = { 0 };
+    TEE_Result res;
+
+    if (ctx == NULL) {
+        return FWK_E_PARAM;
+    }
+
+    if (!is_exposed(ctx)) {
+        /* Return a dummy value to prevent an error trace */
+	*num = 1;
+	*den = 2;
+        return FWK_SUCCESS;
+    }
+
+    res = clk_get_duty_cycle(ctx->clk, &duty);
+
+    FWK_LOG_DEBUG(
+        MOD_NAME "SCMI optee_clock (%u/\"%s\"): %" PRIx32,
+        fwk_id_get_element_idx(dev_id),
+        clk_get_name(ctx->clk),
+        res);
+
+    if (res == TEE_ERROR_NOT_SUPPORTED) {
+        /* Assume a 50% duty cycle */
+        duty = (struct clk_duty){ .num = 1, .den = 2 };
+    } else if (res != TEE_SUCCESS) {
+        return FWK_E_DEVICE;
+    }
+
+    *num = duty.num;
+    *den = duty.den;
+
+    return FWK_SUCCESS;
+}
+
 static int stub_process_power_transition(fwk_id_t dev_id, unsigned int state)
 {
     return FWK_E_SUPPORT;
@@ -325,6 +363,7 @@ static const struct mod_clock_drv_api api_optee_clock = {
     .get_range = get_range,
     .get_rate_from_index = get_rate_from_index,
     .set_rate = set_rate,
+    .get_duty_cycle = get_duty_cycle,
     /* Not supported */
     .process_power_transition = stub_process_power_transition,
     .process_pending_power_transition = stub_pending_power_transition,
diff --git a/core/lib/scmi-server/SCP-firmware/module/scmi/include/mod_scmi_std.h b/core/lib/scmi-server/SCP-firmware/module/scmi/include/mod_scmi_std.h
index e40734449..c135cfe67 100644
--- a/core/lib/scmi-server/SCP-firmware/module/scmi/include/mod_scmi_std.h
+++ b/core/lib/scmi-server/SCP-firmware/module/scmi/include/mod_scmi_std.h
@@ -164,6 +164,7 @@ enum scmi_clock_command_id {
     MOD_SCMI_CLOCK_NAME_GET = 0x008,
     MOD_SCMI_CLOCK_RATE_NOTIFY = 0x009,
     MOD_SCMI_CLOCK_RATE_CHANGE_REQUESTED_NOTIFY = 0x00A,
+    MOD_SCMI_CLOCK_DUTY_CYCLE_GET = 0x00B,
     MOD_SCMI_CLOCK_COMMAND_COUNT
 };
 
diff --git a/core/lib/scmi-server/SCP-firmware/module/scmi_clock/include/internal/scmi_clock.h b/core/lib/scmi-server/SCP-firmware/module/scmi_clock/include/internal/scmi_clock.h
index 3bf65bf6c..3de3ac072 100644
--- a/core/lib/scmi-server/SCP-firmware/module/scmi_clock/include/internal/scmi_clock.h
+++ b/core/lib/scmi-server/SCP-firmware/module/scmi_clock/include/internal/scmi_clock.h
@@ -286,4 +286,18 @@ struct scmi_clock_rate_change_request_notify_p2a {
     int32_t status;
 };
 
+/*
+ * Clock get duty cycle
+ */
+
+struct scmi_clock_duty_cycle_get_a2p {
+    uint32_t clock_id;
+};
+
+struct scmi_clock_duty_cycle_get_p2a {
+    int32_t status;
+    uint32_t numerator;
+    uint32_t denominator;
+};
+
 #endif /* INTERNAL_SCMI_CLOCK_H */
diff --git a/core/lib/scmi-server/SCP-firmware/module/scmi_clock/src/mod_scmi_clock.c b/core/lib/scmi-server/SCP-firmware/module/scmi_clock/src/mod_scmi_clock.c
index f54361a7a..0508d8bdd 100644
--- a/core/lib/scmi-server/SCP-firmware/module/scmi_clock/src/mod_scmi_clock.c
+++ b/core/lib/scmi-server/SCP-firmware/module/scmi_clock/src/mod_scmi_clock.c
@@ -134,6 +134,8 @@ static int scmi_clock_rate_notify_handler(fwk_id_t service_id,
     const uint32_t *payload);
 static int scmi_clock_rate_change_request_notify_handler(fwk_id_t service_id,
     const uint32_t *payload);
+static int scmi_clock_duty_cycle_get_handler(fwk_id_t service_id,
+    const uint32_t *payload);
 
 /*
  * Internal variables.
@@ -155,6 +157,7 @@ static int (*const handler_table[MOD_SCMI_CLOCK_COMMAND_COUNT])(
     [MOD_SCMI_CLOCK_NAME_GET] = scmi_clock_name_get_handler,
     [MOD_SCMI_CLOCK_RATE_NOTIFY] = scmi_clock_rate_notify_handler,
     [MOD_SCMI_CLOCK_RATE_CHANGE_REQUESTED_NOTIFY] = scmi_clock_rate_change_request_notify_handler,
+    [MOD_SCMI_CLOCK_DUTY_CYCLE_GET] = scmi_clock_duty_cycle_get_handler,
 };
 
 static const unsigned int payload_size_table[MOD_SCMI_CLOCK_COMMAND_COUNT] = {
@@ -178,6 +181,8 @@ static const unsigned int payload_size_table[MOD_SCMI_CLOCK_COMMAND_COUNT] = {
         (unsigned int)sizeof(struct scmi_clock_rate_notify_a2p),
     [MOD_SCMI_CLOCK_RATE_CHANGE_REQUESTED_NOTIFY] =
         (unsigned int)sizeof(struct scmi_clock_rate_change_request_notify_a2p),
+    [MOD_SCMI_CLOCK_DUTY_CYCLE_GET] =
+        (unsigned int)sizeof(struct scmi_clock_duty_cycle_get_a2p),
 };
 
 /*
@@ -1474,6 +1479,62 @@ static int scmi_clock_rate_change_request_notify_handler(fwk_id_t service_id,
     return FWK_SUCCESS;
 }
 
+static int scmi_clock_duty_cycle_get_handler(fwk_id_t service_id,
+    const uint32_t *payload)
+
+{
+    const struct scmi_clock_duty_cycle_get_a2p *parameters;
+    struct scmi_clock_duty_cycle_get_p2a return_values = {
+        .status = (int32_t)SCMI_SUCCESS
+    };
+    const struct mod_scmi_clock_device *clock_device;
+    unsigned int agent_id;
+    size_t response_size;
+    int status;
+
+    parameters = (const struct scmi_clock_duty_cycle_get_a2p*)payload;
+
+    status = scmi_clock_get_clock_device_entry(
+        service_id, parameters->clock_id, &clock_device);
+    if (status != FWK_SUCCESS) {
+        return_values.status = (int32_t)SCMI_NOT_FOUND;
+        goto exit;
+    }
+
+#ifdef BUILD_HAS_MOD_RESOURCE_PERMS
+    /* Test SET_RATE permission until we have a DUTY_CYCLE permission flag */
+    status = scmi_clock_permissions_handler(
+        parameters->clock_id,
+        service_id,
+        (unsigned int)MOD_SCMI_CLOCK_RATE_SET);
+    if (status != FWK_SUCCESS) {
+        return_values.status = (int32_t)SCMI_DENIED;
+        goto exit;
+    }
+#endif
+
+    status = scmi_clock_ctx.scmi_api->get_agent_id(service_id, &agent_id);
+    if (status != FWK_SUCCESS) {
+        goto exit;
+    }
+
+    status = scmi_clock_ctx.clock_api->get_duty_cycle(clock_device->element_id,
+                                                      &return_values.numerator,
+                                                      &return_values.denominator);
+
+    if (status == FWK_E_SUPPORT) {
+        return_values.status = (int32_t)SCMI_NOT_SUPPORTED;
+    } else if (status != FWK_SUCCESS) {
+        return_values.status = (int32_t)SCMI_GENERIC_ERROR;
+    }
+
+exit:
+    response_size = (return_values.status == SCMI_SUCCESS) ?
+        sizeof(return_values) : sizeof(return_values.status);
+    return scmi_clock_ctx.scmi_api->respond(
+        service_id, &return_values, response_size);
+}
+
 /*
  * SCMI module -> SCMI clock module interface
  */
-- 
2.47.2

